include/xtensor/xhistogram.hpp

```diff
@@ -57,46 +57,94 @@ namespace xt
-        XTENSOR_ASSERT(xt::amin(data)[0] >= bin_edges[0]);
-        XTENSOR_ASSERT(xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1]);
-
-        xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ bin_edges.size() - 1 });
-
-        auto sorter = xt::argsort(data);
-
-        size_type ibin = 0;
+        return histogramImp(std::forward<E1>(data),
+                            std::forward<E2>(data),
+                            std::forward<E3>(weights),
+                            density,
+                            false);
+    }
 
-        for (auto& idx : sorter)
+    namespace detail
+    {
+        /**
+         * @ingroup histogram
+         * @brief Compute the histogram of a set of data.
+         *
+         * @param data The data.
+         * @param bin_edges The bin-edges. It has to be 1-dimensional and monotonic.
+         * @param weights Weight factors corresponding to each data-point.
+         * @param density If true the resulting integral is normalized to 1. [default: false]
+         * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+         */
+        template <class R = double, class E1, class E2, class E3>
+        inline auto histogramImp(E1&& data, E2&& bin_edges, E3&& weights, bool density = false, bool equal_bins=false)
         {
-            while (data[idx] >= bin_edges[ibin + 1] && ibin < bin_edges.size() - 2)
+            using size_type = common_size_type_t<std::decay_t<E1>, std::decay_t<E2>, std::decay_t<E3>>;
+            using value_type = typename std::decay_t<E3>::value_type;
+
+            XTENSOR_ASSERT(data.dimension() == 1);
+            XTENSOR_ASSERT(weights.dimension() == 1);
+            XTENSOR_ASSERT(bin_edges.dimension() == 1);
+            XTENSOR_ASSERT(weights.size() == data.size());
+            XTENSOR_ASSERT(bin_edges.size() >= 2);
+            XTENSOR_ASSERT(std::is_sorted(bin_edges.cbegin(), bin_edges.cend()));
+            XTENSOR_ASSERT(xt::amin(data)[0] >= bin_edges[0]);
+            XTENSOR_ASSERT(xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1]);
+
+            size_t n_bins = bin_edges.size() - 1;
+            xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ n_bins });
+
+            if (equal_bins)
             {
-                ++ibin;
+                std::array<double, 2> bounds = xt::minmax(data)();
+                double left = bounds[0];
+                double right = bounds[1];
+                double norm = 1. / (right - left);
+                for (size_t i = 0; i < data.size(); ++i)
+                {
+                    auto v = static_cast<double>(data(i));
+                    if ( v >= left & v < right )
+                    {
+                        auto i_bin = static_cast<size_t>(static_cast<double>(n_bins) * (data(i) - left) * norm);
+                        count(i_bin) += weights(i);
+                    }
+                    else if ( v == right )
```

Comment:
Since you always have `left <= v <= right`, I think this part of the code can be simplified to:

```cpp
if (v == right)
{
    count(n_bins - 1) += weights(i);
}
else
{
    auto i_bin = static_cast<size_t>(static_cast<double>(n_bins) * (data(i) - left) * norm);
    count(i_bin) += weights(i);
}

Reply:
In `numpy.histogram`, users are allowed to specify `bins` and `range` at the same time. Indeed, this is the main use cases for histogram with equally sized bins. I was thinking if `xtensor` would have this API in the future, then we will need to check the boundary in this function.

Reply:
Ah indeed, thanks for the explanation. Do you want to go one step further and introduce this new API? In both case, I think it could be nice to add some comments above this section so we don't get confused in six months when we read the code again, what do you think?

Reply:
Yes, I can add this new API after this PR. I will also add comments here later.

Reply:
Indeed, I intended to write `auto bounds = xt::minmax(bin_edges)();` instead of `auto bounds = xt::minmax(data)();`. Now it is fixed.